storage/local/frankenstein.go

```diff
@@ -259,3 +266,80 @@ func (i *Ingestor) Collect(ch chan<- prometheus.Metric) {
 	ch <- i.storedChunks
 	ch <- i.chunkStoreFailures
 }
+
+type invertedIndex struct {
+	mtx sync.RWMutex
+	idx map[model.LabelName]map[model.LabelValue][]model.Fingerprint // entries are sorted in fp order?
+}
+
+func newInvertedIndex() *invertedIndex {
+	return &invertedIndex{
+		idx: map[model.LabelName]map[model.LabelValue][]model.Fingerprint{},
+	}
+}
+
+func (i *invertedIndex) add(metric model.Metric, fp model.Fingerprint) {
+	i.mtx.Lock()
+	defer i.mtx.Unlock()
+
+	for name, value := range metric {
+		values, ok := i.idx[name]
+		if !ok {
+			values = map[model.LabelValue][]model.Fingerprint{}
+		}
+		fingerprints, ok := values[value]
+		j := sort.Search(len(fingerprints), func(i int) bool {
+			return fingerprints[i] >= fp
+		})
+		fingerprints = append(fingerprints, 0)
+		copy(fingerprints[j+1:], fingerprints[j:])
+		fingerprints[j] = fp
+		values[value] = fingerprints
+		i.idx[name] = values
+	}
+}
+
+func (i *invertedIndex) lookup(matchers []*metric.LabelMatcher) []model.Fingerprint {
+	if len(matchers) == 0 {
+		return nil
+	}
+	i.mtx.RLock()
+	defer i.mtx.RUnlock()
+
+	// intersection is initially nil, which is a special case.
+	var intersection []model.Fingerprint
+	for _, matcher := range matchers {
+		values, ok := i.idx[matcher.Name]
+		if !ok {
+			return nil
+		}
+		for value, fps := range values {
+			if matcher.Match(value) {
+				intersection = intersect(intersection, fps)
```

Comment:
I think the intersection needs to happen between label matchers, not between sets of fingerprints resulting from the same label matcher.

For example, `{foo=~"ba."}` can match the label values `bar` and `baz` here, yielding two sets of fingerprints which should be joined, not intersected.

In this line here, you can simply append to a list of fingerprints for the entire matcher, not worrying about duplicate fps, as we know the series have different values for the same label and can't have the same fingerprint. And then intersect with the results of the other matchers.